components/script/dom/offscreencanvas.rs

```diff
@@ -0,0 +1,156 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+
+use dom::bindings::codegen::Bindings::OffscreenCanvasBinding;
+use dom::bindings::error::{Error, Fallible, report_pending_exception};
+use dom::bindings::codegen::Bindings::OffscreenCanvasBinding::OffscreenCanvasMethods;
+use dom::webglrenderingcontext::{LayoutCanvasWebGLRenderingContextHelpers, WebGLRenderingContext};
+use dom::offscreencanvasrenderingcontext2d::{OffscreenCanvasRenderingContext2D, LayoutOffscreenCanvasRenderingContext2DHelpers};
+use dom::htmlelement::HTMLElement;
+use dom::globalscope::GlobalScope;
+use dom::element::{Element, RawLayoutElementHelpers};
+use dom::bindings::cell::DomRefCell;
+use dom::bindings::reflector::{Reflector, reflect_dom_object};
+use dom::bindings::root::{Dom, DomRoot, LayoutDom};
+use dom::bindings::inheritance::Castable;
+use script_layout_interface::{HTMLCanvasData, HTMLCanvasDataSource};
+use dom::node::{Node, window_from_node};
+use html5ever::{LocalName, Prefix};
+use dom::bindings::root::Root;
+use dom::document::Document;
+use style::attr::{AttrValue, LengthOrPercentageOrAuto};
+
+
+use dom_struct::dom_struct;
+
+const DEFAULT_WIDTH: u32 = 300;
+const DEFAULT_HEIGHT: u32 = 150;
+
+#[must_root]
+#[derive(Clone, JSTraceable, MallocSizeOf)]
+pub enum CanvasContext {
+    Context2d(Dom<OffscreenCanvasRenderingContext2D>),
+    WebGL(Dom<WebGLRenderingContext>),
+}
+
+pub trait LayoutOffscreenCanvasHelpers {
+    fn data(&self) -> HTMLCanvasData;
+//	    fn get_width(&self) -> LengthOrPercentageOrAuto;
+//	    fn get_height(&self) -> LengthOrPercentageOrAuto;
+
+	}
+
+	#[dom_struct]
+	pub struct OffscreenCanvas {
+	    htmlelement: HTMLElement,
+	    context: DomRefCell<Option<CanvasContext>>,
+	}
+
+	impl OffscreenCanvas {
+
+	    fn new_inherited(local_name: LocalName,
+			     prefix: Option<Prefix>,
+			     document: &Document) -> OffscreenCanvas {
+		OffscreenCanvas {
+		    htmlelement: HTMLElement::new_inherited(local_name, prefix, document),
+		    context: DomRefCell::new(None),
+		}
+    }
+
+
+    pub fn Constructor(global : &GlobalScope, width: u64,
+                        height: u64) -> Result<DomRoot<OffscreenCanvas>, Error> {
+
+        unimplemented!()
```

Comment:
This method should return a new instance of OffscreenCanvas that is set up according to https://html.spec.whatwg.org/multipage/canvas.html#dom-offscreencanvas .

Reply:
Is it something like below:
pub fn Constructor(global : &GlobalScope, width: u64,
			height: u64) -> Result<DomRoot<OffscreenCanvas>, Error> {
            let context = match canvas.get_or_init_2d_context() {
                Some(context) => context,
                None => return Err(Error::InvalidState),
            };		 
		 Ok(OffscreenCanvas::new(global, x, y, width, height))
}

Reply:
Yes, except:
* that code desn't use context (so no need for it)
* where are x and y coming from? what are they for?

Reply:
Ohh, yeah. so first we need to set the attributes (width and height) to the OffscreenCanvas object and then pass a single parameter -"global" to the fn new that looks like:

Ok(OffscreenCanvas::new(global))

Reply:
No; OffscreenCanvas::new is the method that invokes OffscreenCanvas::new_inherited, which is the method that actually creates an instance of OffscreenCanvas. Traditionally we pass in to new_inherited the initial values that can be controlled by arguments to the constructor.